### DISCUSSION TITLE
How to use maturin with hatch ?

### DISCUSSION THREAD

**Discussion by patrickelectric** [PyO3/maturin#1714] [Q&A] [answered]
I'm using [hatch](https://github.com/pypa/hatch) in my project that uses maturin to generate rust binds to python. I have it working with pyproject using maturin as `build-backend`, but with hatch I need to set build-backend to hatchling.build.

My best regards,

**Comment by messense** [comment]
I think hatch and hatchling are different things, hatch is for package management while hatchling is a PEP 517 build backend just like maturin.

My understanding is that hatch should work with any PEP 517 build backend not tied to hatchling.build. So please provide more information if you are having issue with hatch + maturin.

**Reply by patrickelectric** [reply]
hatch requires using hatching as build-backend, and that's the main problem where I can't use maturin for bind generation.

**Reply by messense** [reply]
I've never used hatch/hatchling before so I don't really know, maybe @ofek can shed some light.

**Reply by ofek** [reply]
Hey! In the next release the `hatch build` command will be able to build packages with other backends, other things like environment management are already agnostic. Use a different build command for now like `python -m build`

**Reply by patrickelectric** [reply]
thanks! Is there a release milestone track or something to help me identify when the release will be available ?

**Reply by ofek** [reply]
You can watch the repo for the next release. I'm currently swamped with work and PEP 723 and in the middle of adding Python management to Hatch

**Reply by patrickelectric** [reply]
Awesome, thanks for you work @ofek!

**Comment by RaulTrombin** [comment]
Some update on this ? Hatch seems way more cleanner to organize the project and automate the building tasks for maturin. Just missing this. It's working pretty fine with pyo3+sphinx

**Reply by RaulTrombin** [reply]
Hi @ofek ! How are doing?

Do you know if it's already implemented?

**Reply by ofek** [reply]
Are you referring to this? https://github.com/PyO3/maturin/discussions/1714#discussioncomment-6658961

**Reply by RaulTrombin** [reply]
yes!

**Reply by RaulTrombin** [reply]
hatch seems way more easy to use and change btwn workspaces, would like to use it

**Reply by ofek** [reply, marked as answer]
Will be released by mid-December https://github.com/pypa/hatch/pull/1018

**Reply by RaulTrombin** [reply]
thank you @ofek! Looking forward to implementing it on my CIs

**Comment by manmartgarc** [comment]
Quick follow up on this, I'm trying to use `maturin develop` but not getting any luck executing this from the hatch environment. Is this because it's a build tool and not a direct dependency?

**Reply by ofek** [reply]
Can you please explain what you want to do and what you have tried so far?

**Reply by manmartgarc** [reply]
I have a hatch package with maturin build system and I can only build it via hatch build. I tried launching the venv with hatch shell but maturin is not available as an executable in the venv so I cannot run maturin develop. I'm guessing this is because maturin is the build system and not a direct dependency.

**Reply by ofek** [reply]
Yes that is the issue, build environments are managed separately and are temporary when using a backend other than Hatchling. The next minor release they will no longer be temporary for other backends as they are now.

**Reply by manmartgarc** [reply]
Cool, thanks for the quick reply 👍

**Reply by flying-sheep** [reply]
Hmm, the problem is that I can’t tell Hatch to replace installation with another command:

```toml
[envs.hatch-test]
pre-install-commands = ["maturin develop --uv"]
```

Kust means that it runs that (possibly skipping compile), and then Hatch runs the maturin build backend which compiles the package a second time.

`skip-install = true` skips `pre-install-commands` as well.
